plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/PingTask.java

```diff
@@ -0,0 +1,144 @@
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TimerTask;
+import java.util.stream.Collectors;
+
+public class PingTask extends TimerTask {
+
+    private static final String MONITORABLE_SERVICE_NAME = "monitorable";
+
+    private static final Logger logger = LoggerFactory.getLogger(PingTask.class);
+
+    private static final String PING_PATH = "/ping";
+
+    private final ServiceQuery serviceQuery;
+    private final HttpClient httpClient;
+    private final AlarmManager alarmManager;
+    private final PlantDescriptionTracker pdTracker;
+
+    public PingTask(
+        final ServiceQuery serviceQuery,
+        final HttpClient httpClient,
+        final AlarmManager alarmManager,
+        PlantDescriptionTracker pdTracker
+    ) {
+
+        Objects.requireNonNull(serviceQuery, "Expected service query");
+        Objects.requireNonNull(httpClient, "Expected HTTP client");
+        Objects.requireNonNull(alarmManager, "Expected alarm manager");
+        Objects.requireNonNull(pdTracker, "Expected Plant Description tracker");
+
+        this.serviceQuery = serviceQuery;
+        this.httpClient = httpClient;
+        this.alarmManager = alarmManager;
+        this.pdTracker = pdTracker;
+    }
+
+    @Override
+    public void run() {
+
+        serviceQuery.resolveAll()
+            .ifSuccess(services -> {
+                updateAlarmsForMissingServices(services);
+                services.forEach(this::ping);
+            })
+            .onFailure(e -> logger.error("Failed to ping monitorable systems.", e));
+    }
+
+    private void updateAlarmsForMissingServices(Set<ServiceRecord> detectedMonitorableServices) {
+
+        // TODO: Maybe this should be done per expected monitorable service
+        // (i.e. per port with service definition 'monitorable'), not per
+        // system?
+        List<PdeSystem> expectedMonitorableSystems = getMonitorableSystems();
+        for (PdeSystem system : expectedMonitorableSystems) {
+
+            boolean hasMonitorableService = detectedMonitorableServices.stream()
+                .anyMatch(service -> isProvidedBy(service.provider(), system));
+
+            if (hasMonitorableService) {
+                alarmManager.clearSystemNotMonitorable(
+                    system.systemId(),
+                    system.systemName().orElse(null),
+                    system.metadata()
+                );
+            } else {
+                alarmManager.raiseSystemNotMonitorable(
+                    system.systemId(),
+                    system.systemName().orElse(null),
+                    system.metadata()
+                );
+            }
+        }
+    }
+
+    private boolean isProvidedBy(SystemRecord provider, PdeSystem system) {
+        boolean metadataMatch = Metadata.isSubset(system.metadata(), provider.metadata());
+        boolean nameMatch = system.systemName().isEmpty() || system.systemName().get().equals(provider.name());
```

Comment:
possible issues because of system name normalization (if any name comes from the SR)

Reply:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.